mbus-api/src/v0.rs

```diff
@@ -182,3 +202,276 @@ pub struct Node {
 #[derive(Serialize, Deserialize, Default, Debug, Clone)]
 pub struct Nodes(pub Vec<Node>);
 bus_impl_message_all!(GetNodes, GetNodes, Nodes, Node);
+
+/// Pool Service
+/// Filter query objects
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum Filter {
+    /// All objects
+    All,
+    /// Filter by Node id
+    Node(String),
+    /// Pool filters
+    ///
+    /// Filter by Pool id
+    Pool(String),
+    /// Filter by Node and Pool id
+    NodePool(String, String),
+    /// Filter by Node and Replica id
+    NodeReplica(String, String),
+    /// Filter by Node, Pool and Replica id
+    NodePoolReplica(String, String, String),
```

Comment:
These filters are cool, but I just wonder how easy it is to for example swap to strings? In a different commit I see we do lots of type definitions (good) but why don't we use it? i.e

struct NodeReplica { NodeName, PoolName, ReplicaName }, ?

Reply:
You mean so it'd become like:
```
let replicas = get_replicas(Filter::NodePoolReplica {
node_name: "node".into(),
pool_name: "pool".into(),
replica_id: "1234...".into()
}).await?;
```

Reply:
Kinda sort yes. This way, when writing the code, and due to the rich nature of the rust tooling, you can by accident replace the node_name string, with a let's say, pool_name. 

```rust
let node_name  = NodeName::from(arg1);
let pool_name = PoolName::from(arg2);
let replica_id = ReplicaId::from(arg3);

let replica = get_replicas( Filter::NoneRepica { node_name, pool_name, replica_id }).await?;
```

This also will allow for, in the future nicer builder patterns? 

```rust
let replicas = Filter.with_nodes(vec![arg1]).with_pool("pool").build().await?;
```

This is a lot of personal taste of course and I'm not suggesting that the above is better or is nicer or whatever, but:

 - as arguments grow, builder patterns are nice
 - we have strongly typed system lets try to make use of it!
 - bikeshedding now, about these horrible small details, will prevent redoing things later

Reply:
I like the builder patter there! Maybe the `T` can be inferred from usage, so you could have all nexus from the node, or all pools from the node, like so:
```
let nexus: Vec<Nexus> = Filter.with_node(MY_NODE).build().await?;
let pools: Vec<Pool> = Filter.with_node(MY_NODE).build().await?;
```
But I'll probably start with the type changes first and leave the builder for latter.

Reply:
@gila coming back to this, would adding a specific type to each (NodeName/ReplicaId/etc) suffice? And keep the tuple since now you cannot swap the types?:
`// this makes it very explicit
let node_name  = NodeName::from(arg1);
let pool_name = PoolName::from(arg2);
let replica_id = ReplicaId::from(arg3);

// type checker will catch the error here
let replica = get_replicas(node_name, pool_name, replica_id).await?;`

Reply:
Yes I think so! A tuple (roughly to my understanding) just is a struct with nameless members so that would work just fine!